WcaOnRails/lib/advancement_condition.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+class AdvancementCondition
+  include ActiveModel::Validations
+
+  attr_accessor :level
+  validates :level, numericality: { only_integer: true }
+
+  def initialize(level)
+    self.level = level
+  end
+
+  def to_wcif
+    { "type" => self.class.wcif_type, "level" => self.level }
+  end
+
+  def ==(other)
+    other.class == self.class && other.to_wcif == self.to_wcif
+  end
+
+  def hash
+    self.to_wcif.hash
+  end
+
+  def self.wcif_type_to_class
+    @@wcif_type_to_class ||= AdvancementCondition.subclasses.map { |cls| [cls.wcif_type, cls] }.to_h
```

Comment:
There was a littleb it of other code dealing with this mapping! However, we updated the WCIF spec for `AdvancmentCondition`, which had a nice side effect of making this code simpler.

## The old definition

```
{
  "type": "ranking"|"percentile"|"attemptValue",
  "ranking": Integer,          // Only used if "type" == "ranking".
  "percentile": Integer,       // Only used if "type" == "percentile".
  "attemptValue": AttemptValue // Only used if "type" == "attemptValue".
}
```

## The new definition

```
{
  "type": "ranking"|"percent"|"attemptResult",
  "level": Ranking|Percent|AttemptResult
}
```

You can imagine how with the new definition, it's simpler to unconditionally read and write `level`, rather than determining what the key name should be.